src/analyses/constant_propagator.cpp

```diff
@@ -510,7 +514,46 @@ bool constant_propagator_domaint::try_evaluate(
   exprt &expr,
   const namespacet &ns) const
 {
+  const irep_idt ID_rounding_mode = CPROVER_PREFIX "rounding_mode";
   adjust_float_expressions(expr, ns);
+
+  // if the current rounding mode is top we can
+  // still get a non-top result by trying all rounding
+  // modes and checking if the results are all the same
+  if(!values.is_constant(symbol_exprt(ID_rounding_mode)))
+  {
+    // NOLINTNEXTLINE (whitespace/braces)
+    auto rounding_modes = std::array<ieee_floatt::rounding_modet, 4>{
+      // NOLINTNEXTLINE (whitespace/braces)
+      {ieee_floatt::ROUND_TO_EVEN,
+       ieee_floatt::ROUND_TO_ZERO,
+       ieee_floatt::ROUND_TO_MINUS_INF,
+       // NOLINTNEXTLINE (whitespace/braces)
+       ieee_floatt::ROUND_TO_PLUS_INF}};
+    // instead of 4, we could write rounding_modes.size() here
+    // if we dropped Visual Studio 2013 support (no constexpr)
+    std::array<exprt, 4> possible_results;
```

Comment:
I don't think this implementation is optimal: there is no need to memorise all results - it suffices to maintain just a single one and you can abort early if the most recently computed is different from that single memorised one.

Reply:
I'm not sure how much of a difference this will make, but you're right there's no need to keep all the results around (this was an attempt to avoid code duplication but the result ended up being more complicated than what I tried to avoid I think), so I will fix this.

Reply:
Arguably this won't make any noticeable performance difference, but simplifying the code is important for maintainability. Thanks!